File: nim/src/jtd.nim
## JSON Type Definition (RFC 8927) validator — mirror of python/jsoncanon/jtd.py.
## Produces standard error indicators (instancePath / schemaPath) as JSON
## Pointers, in a fixed depth-first order, so it is byte-identical to Python.

import std/strutils
import ./jsoncanon

proc jtdGet(o: JNode, key: string): JNode =
  if o.kind != jkObj: return nil
  let i = o.keys.find(key)
  if i >= 0: o.vals[i] else: nil

proc jtdHas(o: JNode, key: string): bool =
  o.kind == jkObj and o.keys.find(key) >= 0

proc isRfc3339(s: string): bool =
  ## ^\d{4}-\d{2}-\d{2}[Tt]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$
  var i = 0
  proc dig(s: string, i: var int, n: int): bool =
    for _ in 0 ..< n:
      if i >= s.len or s[i] notin {'0'..'9'}: return false
      inc i
    true
  proc lit(s: string, i: var int, c: char): bool =
    if i < s.len and s[i] == c: (inc i; true) else: false
  proc litset(s: string, i: var int, cs: set[char]): bool =
    if i < s.len and s[i] in cs: (inc i; true) else: false
  if not dig(s, i, 4) or not lit(s, i, '-') or not dig(s, i, 2) or
     not lit(s, i, '-') or not dig(s, i, 2) or not litset(s, i, {'T', 't'}) or
     not dig(s, i, 2) or not lit(s, i, ':') or not dig(s, i, 2) or
     not lit(s, i, ':') or not dig(s, i, 2): return false
  if i < s.len and s[i] == '.':
    inc i
    var c = 0
    while i < s.len and s[i] in {'0'..'9'}: (inc i; inc c)
    if c == 0: return false
  if litset(s, i, {'Z', 'z'}): discard
  elif i < s.len and s[i] in {'+', '-'}:
    inc i
    if not dig(s, i, 2) or not lit(s, i, ':') or not dig(s, i, 2): return false
  else: return false
  i == s.len

proc isIntIn(numtok: string, lo, hi: int64): bool =
  let plain = canonNumber(numtok, false, nfPlain)
  if '.' in plain: return false
  let digits = if plain[0] == '-': plain[1..^1] else: plain
  if digits.len > 10: return false      # beyond any JTD int range
  let iv = parseBiggestInt(plain)
  iv >= lo and iv <= hi

proc checkType(t: string, inst: JNode): bool =
  case t
  of "boolean": inst.kind == jkBool
  of "string": inst.kind == jkStr
  of "timestamp": inst.kind == jkStr and isRfc3339(inst.str)
  of "float32", "float64": inst.kind == jkNum
  of "int8": inst.kind == jkNum and isIntIn(inst.num, -128, 127)
  of "uint8": inst.kind == jkNum and isIntIn(inst.num, 0, 255)
  of "int16": inst.kind == jkNum and isIntIn(inst.num, -32768, 32767)
  of "uint16": inst.kind == jkNum and isIntIn(inst.num, 0, 65535)
  of "int32": inst.kind == jkNum and isIntIn(inst.num, -2147483648'i64, 2147483647'i64)
  of "uint32": inst.kind == jkNum and isIntIn(inst.num, 0, 4294967295'i64)
  else: false

proc jtdPtr(tokens: seq[string]): string =
  for t in tokens:
    result.add "/"
    result.add t.replace("~", "~0").replace("/", "~1")

proc jtdValidate(root, schema, inst: JNode,
                 errors: var seq[(seq[string], seq[string])],
                 ip, sp: seq[string], parentTag = "", useTag = false) =
  if schema.kind != jkObj: return
  let nul = jtdGet(schema, "nullable")
  if nul != nil and nul.kind == jkBool and nul.b and inst.kind == jkNull: return

  if jtdHas(schema, "ref"):
    let name = jtdGet(schema, "ref").str
    let defs = jtdGet(root, "definitions")
    if defs == nil or not jtdHas(defs, name):
      raise newException(JsonError, "JTD: unknown definition '" & name & "'")
    jtdValidate(root, jtdGet(defs, name), inst, errors, ip, @["definitions", name])
  elif jtdHas(schema, "type"):
    if not checkType(jtdGet(schema, "type").str, inst): errors.add((ip, sp & @["type"]))
  elif jtdHas(schema, "enum"):
    var found = false
    if inst.kind == jkStr:
      for c in jtdGet(schema, "enum").arr:
        if c.kind == jkStr and c.str == inst.str: (found = true; break)
    if not found: errors.add((ip, sp & @["enum"]))
  elif jtdHas(schema, "elements"):
    if inst.kind == jkArr:
      for i, e in inst.arr:
        jtdValidate(root, jtdGet(schema, "elements"), e, errors, ip & @[$i], sp & @["elements"])
    else: errors.add((ip, sp & @["elements"]))
  elif jtdHas(schema, "properties") or jtdHas(schema, "optionalProperties"):
    if inst.kind == jkObj:
      let props = jtdGet(schema, "properties")
      let opts = jtdGet(schema, "optionalProperties")
      if props != nil:
        for k in props.keys:
          if jtdHas(inst, k):
            jtdValidate(root, jtdGet(props, k), jtdGet(inst, k), errors, ip & @[k], sp & @["properties", k])
          else:
            errors.add((ip, sp & @["properties", k]))
      if opts != nil:
        for k in opts.keys:
          if jtdHas(inst, k):
            jtdValidate(root, jtdGet(opts, k), jtdGet(inst, k), errors, ip & @[k], sp & @["optionalProperties", k])
      let addl = jtdGet(schema, "additionalProperties")
      let allowAddl = addl != nil and addl.kind == jkBool and addl.b
      if not allowAddl:
        for k in inst.keys:
          let inProps = props != nil and jtdHas(props, k)
          let inOpts = opts != nil and jtdHas(opts, k)
          if not inProps and not inOpts and not (useTag and k == parentTag):
            errors.add((ip & @[k], sp))
    else:
      errors.add((ip, sp & (if jtdHas(schema, "properties"): @["properties"] else: @["optionalProperties"])))
  elif jtdHas(schema, "values"):
    if inst.kind == jkObj:
      for k in inst.keys:
        jtdValidate(root, jtdGet(schema, "values"), jtdGet(inst, k), errors, ip & @[k], sp & @["values"])
    else: errors.add((ip, sp & @["values"]))
  elif jtdHas(schema, "discriminator"):
    let tag = jtdGet(schema, "discriminator").str
    if inst.kind == jkObj:
      if jtdHas(inst, tag):
        let tv = jtdGet(inst, tag)
        let mapping = jtdGet(schema, "mapping")
        if tv.kind == jkStr and mapping != nil and jtdHas(mapping, tv.str):
          jtdValidate(root, jtdGet(mapping, tv.str), inst, errors, ip, sp & @["mapping", tv.str], tag, true)
        else:
          errors.add((ip & @[tag], sp & @["mapping"]))
      else:
        errors.add((ip, sp & @["discriminator"]))
    else:
      errors.add((ip, sp & @["discriminator"]))
  # empty form (and metadata-only): always valid

proc validateJtd*(schemaRaw, instanceRaw: string, encoding = ""): seq[Issue] =
  let schema = parse(decodeBytes(schemaRaw, encoding))
  let inst = parse(decodeBytes(instanceRaw, encoding))
  var errors: seq[(seq[string], seq[string])]
  jtdValidate(schema, schema, inst, errors, @[], @[])
  for (ip, sp) in errors:
    let il = jtdPtr(ip)
    let sl = jtdPtr(sp)
    result.add(((if il.len > 0: il else: "(root)"), "jtd",
                "schema " & (if sl.len > 0: sl else: "(root)")))

File: python/jsoncanon/jtd.py
"""JSON Type Definition (RFC 8927) validator.

Validates a JSON instance against a JTD schema, producing the standard error
indicators (instancePath / schemaPath) from RFC 8927 §3.3.2, rendered as JSON
Pointers. Implements all eight forms (empty, ref, type, enum, elements,
properties, values, discriminator) plus `nullable` and root `definitions`.
The schema is assumed to be a valid JTD schema.

Deterministic: errors are produced in a fixed depth-first order, so the Nim port
emits an identical list.
"""

from __future__ import annotations

import re

from .lint import Issue
from .numbers import canon_number
from .parser import JValue, Number, decode_bytes, loads, JSONError

_RFC3339 = re.compile(
    r"^\d{4}-\d{2}-\d{2}[Tt]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$")

_INT_RANGES = {
    "int8": (-128, 127), "uint8": (0, 255),
    "int16": (-32768, 32767), "uint16": (0, 65535),
    "int32": (-2147483648, 2147483647), "uint32": (0, 4294967295),
}


def _is_int_in(num: Number, lo: int, hi: int) -> bool:
    plain = canon_number(num.text, number_format="plain")
    if "." in plain:
        return False
    return lo <= int(plain) <= hi


def _check_type(t: str, inst: JValue) -> bool:
    if t == "boolean":
        return inst is True or inst is False
    if t == "string":
        return isinstance(inst, str)
    if t == "timestamp":
        return isinstance(inst, str) and bool(_RFC3339.match(inst))
    if t in ("float32", "float64"):
        return isinstance(inst, Number)
    if t in _INT_RANGES:
        return isinstance(inst, Number) and _is_int_in(inst, *_INT_RANGES[t])
    return False


def _ptr(tokens: list[str]) -> str:
    return "".join("/" + t.replace("~", "~0").replace("/", "~1") for t in tokens)


def _validate(root: JValue, schema: JValue, inst: JValue,
              errors: list[tuple[list[str], list[str]]],
              ip: list[str], sp: list[str], parent_tag: str | None = None) -> None:
    assert isinstance(schema, dict)
    if schema.get("nullable") is True and inst is None:
        return

    if "ref" in schema:
        name = schema["ref"]
        assert isinstance(name, str) and isinstance(root, dict)
        defs = root.get("definitions")
        assert isinstance(defs, dict)
        if name not in defs:
            raise JSONError(f"JTD: unknown definition {name!r}")
        _validate(root, defs[name], inst, errors, ip, ["definitions", name])
    elif "type" in schema:
        t = schema["type"]
        assert isinstance(t, str)
        if not _check_type(t, inst):
            errors.append((ip, sp + ["type"]))
    elif "enum" in schema:
        choices = schema["enum"]
        assert isinstance(choices, list)
        if not (isinstance(inst, str) and inst in choices):
            errors.append((ip, sp + ["enum"]))
    elif "elements" in schema:
        if isinstance(inst, list):
            for i, e in enumerate(inst):
                _validate(root, schema["elements"], e, errors,
                          ip + [str(i)], sp + ["elements"])
        else:
            errors.append((ip, sp + ["elements"]))
    elif "properties" in schema or "optionalProperties" in schema:
        if isinstance(inst, dict):
            props = schema.get("properties") or {}
            opts = schema.get("optionalProperties") or {}
            assert isinstance(props, dict) and isinstance(opts, dict)
            for k in props:
                if k in inst:
                    _validate(root, props[k], inst[k], errors,
                              ip + [k], sp + ["properties", k])
                else:
                    errors.append((ip, sp + ["properties", k]))
            for k in opts:
                if k in inst:
                    _validate(root, opts[k], inst[k], errors,
                              ip + [k], sp + ["optionalProperties", k])
            if schema.get("additionalProperties") is not True:
                for k in inst:
                    if k not in props and k not in opts and k != parent_tag:
                        errors.append((ip + [k], sp))
        else:
            errors.append((ip, sp + (["properties"] if "properties" in schema
                                     else ["optionalProperties"])))
    elif "values" in schema:
        if isinstance(inst, dict):
            for k, v in inst.items():
                _validate(root, schema["values"], v, errors,
                          ip + [k], sp + ["values"])
        else:
            errors.append((ip, sp + ["values"]))
    elif "discriminator" in schema:
        tag = schema["discriminator"]
        assert isinstance(tag, str)
        if isinstance(inst, dict):
            if tag in inst:
                tv = inst[tag]
                mapping = schema["mapping"]
                assert isinstance(mapping, dict)
                if isinstance(tv, str) and tv in mapping:
                    _validate(root, mapping[tv], inst, errors, ip,
                              sp + ["mapping", tv], parent_tag=tag)
                else:
                    errors.append((ip + [tag], sp + ["mapping"]))
            else:
                errors.append((ip, sp + ["discriminator"]))
        else:
            errors.append((ip, sp + ["discriminator"]))
    # empty form (and metadata-only): always valid


def validate_jtd(schema_raw: bytes, instance_raw: bytes, *,
                 encoding: str | None = None) -> list[Issue]:
    schema = loads(decode_bytes(schema_raw, encoding))
    inst = loads(decode_bytes(instance_raw, encoding))
    errors: list[tuple[list[str], list[str]]] = []
    _validate(schema, schema, inst, errors, [], [])
    return [Issue(_ptr(ip) or "(root)", "jtd", "schema " + (_ptr(spp) or "(root)"))
            for ip, spp in errors]
